compositor_render/src/transformations/image_renderer.rs

```diff
@@ -233,6 +278,140 @@ impl SvgAsset {
     }
 }
 
+pub struct AnimatedNodeState {
+    first_pts: Option<Duration>,
+}
+
+pub struct AnimatedAsset {
+    frames: Vec<AnimationFrame>,
+    animation_duration: Duration,
+}
+
+struct AnimationFrame {
+    texture: RGBATexture,
+    pts: Duration,
+}
+
+impl AnimatedAsset {
+    fn new(ctx: &WgpuCtx, data: Bytes, format: ImageFormat) -> Result<Self, AnimatedError> {
+        let decoded_frames = match format {
+            ImageFormat::Gif => GifDecoder::new(&data[..])?.into_frames(),
+            _ => todo!(),
+        };
+
+        let mut animation_duration: Duration = Duration::ZERO;
+        let mut frames = vec![];
+        for frame in decoded_frames {
+            let frame = &frame?;
+            let buffer = frame.buffer();
+            let texture = RGBATexture::new(
+                ctx,
+                Resolution {
+                    width: buffer.width() as usize,
+                    height: buffer.height() as usize,
+                },
+            );
+            texture.upload(ctx, buffer);
+
+            let delay: Duration = frame.delay().into();
+            animation_duration += delay;
+            frames.push(AnimationFrame {
+                texture,
+                pts: animation_duration,
+            });
+
+            if frames.len() > 1000 {
+                return Err(AnimatedError::TooMuchFrames);
+            }
+        }
+
+        let Some(first_frame) = frames.first() else {
+            return Err(AnimatedError::NoFrames)
+        };
+        if frames.len() == 1 {
+            return Err(AnimatedError::SingleFrame);
+        }
+        let first_frame_size = first_frame.texture.size();
+        if !frames
+            .iter()
+            .all(|frame| frame.texture.size() == first_frame_size)
+        {
+            return Err(AnimatedError::UnsupportedVariableResolution);
+        }
+
+        ctx.queue.submit([]);
+
+        // In case only one frame, where first delay is zero
+        if animation_duration.is_zero() {
+            animation_duration = Duration::from_nanos(1)
+        }
+
+        Ok(Self {
+            frames,
+            animation_duration,
+        })
+    }
+
+    fn render(
+        &self,
+        ctx: &WgpuCtx,
+        target: &NodeTexture,
+        state: &Mutex<AnimatedNodeState>,
+        pts: Duration,
+    ) {
+        let mut state = state.lock().unwrap();
+        let first_pts = match state.first_pts {
+            Some(first_pts) => first_pts,
+            None => {
+                state.first_pts = Some(pts);
+                pts
+            }
+        };
+
+        let animation_pts = Duration::from_nanos(
+            ((pts.as_nanos() - first_pts.as_nanos()) % self.animation_duration.as_nanos()) as u64,
+        );
+
+        let closest_frame = self
+            .frames
+            .iter()
+            .min_by_key(|frame| u128::abs_diff(frame.pts.as_nanos(), animation_pts.as_nanos()))
+            .unwrap();
```

Comment:
[NIT]
For n frames in animated assets, this takes O(n^2) for all frames (calling `render` n times). Since `frames` is already sorted, you can just use bin search to bring it to O(nlogn) or use BTreeMap. With a stateful iterator for each node and sorted vec, you can get it down O(n) (using sth similar to cyclic buffer).
IDK how much important it's to optimize it, but you can consider it since this runs on every render call

Reply:
You already have `AnimatedNodeState`, so it's quite simple to use optimal iter here

Reply:
> For n frames in animated assets, this takes O(n^2) for all frames (calling render n times).

but we are not calling render n time, so this O(n), but you are right this could easilly be O(log n) and if we want to even constant.

My reasoning was that we have a hard limit on frames and the creation of frames is relatively expensive, so I treated the number of frames as constant

Overall I think the performance difference is negligible (Iterating over list of up to 1000 elements is nothing).I don't want to implement my own binary search or custom iterator, because it would affect readability, but I'll try using BTreeMap

Reply:
I would stick with naive iteration here. The time this would take is probably too short to notice compared to uploading a frame to the GPU.

Reply:
and if this was a performance bottleneck, it's easy to notice through profiling and fix

Reply:
oh, I just noticed you changed it already. nevermind then.

Reply:
yes, after implementing we decided that increased complexity is not worth it